File: streamProcessors/mongoShellExamples/createCustomersASP.js
// This stream processor captures changes from the 'customers' collection
// and merges customer details into the 'customerWithAllDataASP' collection.
// It transforms the data by concatenating the customer first and last names and elevates certain fields to the root of the document.

source = {
      "$source": {
        "coll": "customers",
        "connectionName": "AtlasDatabase",
        "db": "dealer_sourceDB"
      }
    }

m = {
    "$match": {
      "operationType": "insert"
    }
  }
p = {
        "$project": {
            "_id": "$fullDocument._id",
            "name": { "$concat": ["$fullDocument.firstName", " ", "$fullDocument.lastName"] },
            "customerDetails.firstName": "$fullDocument.firstName",
            "customerDetails.lastName": "$fullDocument.lastName",
            "customerDetails.email": "$fullDocument.email",
            "customerDetails.phone": "$fullDocument.phone",
            "CustomerCreatedAt": "$fullDocument.createdAt",
            "isDeleted": "$fullDocument.isDeleted",
            "CustomerLastModified": "$fullDocument.lastModified"
        }
    }
merge = {
      "$merge": {
        "into": {
            "coll": "customerWithAllDataASP",
            "connectionName": "AtlasDatabase",
            "db": "dealer_targetDB"
        }
    }
}

deadLetter = {
  dlq: {
    connectionName: "AtlasDatabase",
    db: "ASPDeadLetterDB",
    coll: "dlq"
  }
}
print("Dropping 'customers' Stream Processor...")
if (sp.listStreamProcessors({"name":"customers"}).length != 0) {sp.customers.drop()}
print("Creating 'customers' Stream Processor...")
sp.createStreamProcessor("customers", [source, m, p, merge], deadLetter)
print("Starting 'customers' Stream Processor...")
sp.customers.start()

File: streamProcessors/restAPIExamples/createCustomersASP.py
from streamProcessorsHelper import create_sp, start_sp, delete_sp
import configparser

# MongoDB Atlas API configuration
config = configparser.ConfigParser()
config.read("config.ini")

WORKSPACE_NAME = config.get("DEFAULT", "WORKSPACE_NAME")

source = {
      "$source": {
        "coll": "customerAgreements",
        "connectionName": "AtlasDatabase",
        "db": "dealer_sourceDB"
      }
    }

m = {
    "$match": {
      "operationType": "insert"
    }
  }

p = {
        "$project": {
          "_id": "$fullDocument.customerID", 
          "customerAgreement": {"$concatArrays": [["$fullDocument"], []]},
          "latestAgreement": "$fullDocument"
        }
    }

merge = {
      "$merge": {
        "into": {
            "coll": "customerWithAllDataASP",
            "connectionName": "AtlasDatabase",
            "db": "dealer_targetDB"
        },
        "whenMatched": [
          {
            "$set": {
              "customerAgreement" : { "$cond":
                {
                  "if": {"$isArray": "$customerAgreement" },
                  "then": { "$concatArrays": [ "$customerAgreement", "$$new.customerAgreement" ] },
                  "else": {"$concatArrays": ["$$new.customerAgreement",[]]}
                }
              },
              "latestAgreement": "$$new.latestAgreement"
            }
          }
        ]
    }
}

deadLetter = {
  "dlq": {
    "connectionName": "AtlasDatabase",
    "db": "ASPDeadLetterDB",
    "coll": "dlq"
  }, "tier": "SP2"
}

SP_CONFIG = {
    "name": "customerAgreements",
    "pipeline": [source, m, p, merge],
    "options": deadLetter
}


def main():
    print(f"Deleting existing stream processor {SP_CONFIG['name']} if it exists...")
    delete_sp(WORKSPACE_NAME, SP_CONFIG['name'])
    print(f"Creating stream processing for {SP_CONFIG['name']}...")
    create_sp(WORKSPACE_NAME, SP_CONFIG)
    print(f"Starting stream processor {SP_CONFIG['name']}...")
    start_sp(WORKSPACE_NAME, SP_CONFIG['name'])

if __name__ == "__main__":
    main()  